File: data/coingecko.py
"""CoinGecko API client. Free tier, no key required.

Rate limit: ~30 calls/min on the public endpoint. All calls go through
st.cache_data with generous TTLs to stay well under that budget.
"""
from __future__ import annotations

import time
from typing import Any

import pandas as pd
import requests
import streamlit as st

BASE_URL = "https://api.coingecko.com/api/v3"
TIMEOUT = 15

CORE_COINS = [
    {"id": "bitcoin", "symbol": "btc", "name": "Bitcoin",
     "tv": "BINANCE:BTCUSDT", "tag": "Store of Value"},
    {"id": "ethereum", "symbol": "eth", "name": "Ethereum",
     "tv": "BINANCE:ETHUSDT", "tag": "Smart Contracts"},
    {"id": "solana", "symbol": "sol", "name": "Solana",
     "tv": "BINANCE:SOLUSDT", "tag": "Layer 1"},
]


def _get(path: str, params: dict | None = None) -> Any:
    url = f"{BASE_URL}{path}"
    for attempt in range(3):
        resp = requests.get(url, params=params, timeout=TIMEOUT)
        if resp.status_code == 429:
            time.sleep(5 * (attempt + 1))
            continue
        resp.raise_for_status()
        return resp.json()
    resp.raise_for_status()
    return resp.json()


@st.cache_data(ttl=60, show_spinner=False)
def get_markets(coin_ids: list[str], vs_currency: str = "usd") -> list[dict]:
    """Current market snapshot for the given coin ids."""
    if not coin_ids:
        return []
    data = _get(
        "/coins/markets",
        {
            "vs_currency": vs_currency,
            "ids": ",".join(coin_ids),
            "order": "market_cap_desc",
            "per_page": len(coin_ids),
            "page": 1,
            "sparkline": "true",
            "price_change_percentage": "1h,24h,7d,30d",
        },
    )
    return data


@st.cache_data(ttl=300, show_spinner=False)
def get_ohlc(coin_id: str, days: int = 30, vs_currency: str = "usd") -> pd.DataFrame:
    """OHLC candles. CoinGecko granularity: 1-2 → 30m, 3-30 → 4h, 31+ → 4d."""
    data = _get(
        f"/coins/{coin_id}/ohlc",
        {"vs_currency": vs_currency, "days": days},
    )
    if not data:
        return pd.DataFrame(columns=["time", "open", "high", "low", "close"])
    df = pd.DataFrame(data, columns=["time", "open", "high", "low", "close"])
    df["time"] = pd.to_datetime(df["time"], unit="ms")
    return df


@st.cache_data(ttl=900, show_spinner=False)
def get_market_chart(coin_id: str, days: int = 90, vs_currency: str = "usd") -> pd.DataFrame:
    """Historical price + volume. Returns empty DataFrame on any error."""
    _empty = pd.DataFrame(columns=["time", "close", "volume"])
    try:
        data = _get(
            f"/coins/{coin_id}/market_chart",
            {"vs_currency": vs_currency, "days": days},
        )
    except Exception:
        return _empty
    prices = pd.DataFrame(data.get("prices", []), columns=["time", "close"])
    volumes = pd.DataFrame(data.get("total_volumes", []), columns=["time", "volume"])
    if prices.empty:
        return _empty
    df = prices.merge(volumes, on="time", how="left")
    df["time"] = pd.to_datetime(df["time"], unit="ms")
    return df


@st.cache_data(ttl=600, show_spinner=False)
def search_coins(query: str) -> list[dict]:
    if not query or len(query.strip()) < 2:
        return []
    data = _get("/search", {"query": query.strip()})
    return data.get("coins", [])[:15]

File: lib/providers/coingecko.ts
import "server-only";

const BASE = "https://api.coingecko.com/api/v3";

export const COIN_MAP: Record<string, { id: string; name: string; symbol: string; pair: string }> = {
  btc:  { id: "bitcoin",      name: "Bitcoin",   symbol: "BTC",  pair: "Bitcoin / US Dollar" },
  eth:  { id: "ethereum",     name: "Ethereum",  symbol: "ETH",  pair: "Ethereum / US Dollar" },
  sol:  { id: "solana",       name: "Solana",    symbol: "SOL",  pair: "Solana / US Dollar" },
  bnb:  { id: "binancecoin",  name: "BNB",       symbol: "BNB",  pair: "BNB / US Dollar" },
  xrp:  { id: "ripple",       name: "XRP",       symbol: "XRP",  pair: "XRP / US Dollar" },
  doge: { id: "dogecoin",     name: "Dogecoin",  symbol: "DOGE", pair: "Dogecoin / US Dollar" },
  ada:  { id: "cardano",      name: "Cardano",   symbol: "ADA",  pair: "Cardano / US Dollar" },
};

export type MarketRow = {
  id: string;
  symbol: string;
  name: string;
  image: string;
  current_price: number;
  market_cap: number;
  price_change_percentage_1h_in_currency?: number;
  price_change_percentage_24h_in_currency?: number;
  price_change_percentage_7d_in_currency?: number;
  price_change_percentage_30d_in_currency?: number;
  sparkline_in_7d?: { price: number[] };
  total_volume: number;
};

async function getJson<T>(url: string, revalidate: number): Promise<T | null> {
  try {
    const res = await fetch(url, { next: { revalidate } });
    if (!res.ok) return null;
    return (await res.json()) as T;
  } catch {
    return null;
  }
}

export async function getMarkets(slugs: string[]): Promise<MarketRow[]> {
  const ids = slugs.map((s) => COIN_MAP[s]?.id).filter(Boolean).join(",");
  if (!ids) return [];
  const url =
    `${BASE}/coins/markets?vs_currency=usd&ids=${ids}` +
    `&order=market_cap_desc&per_page=${slugs.length}&page=1` +
    `&sparkline=true&price_change_percentage=1h,24h,7d,30d`;
  return (await getJson<MarketRow[]>(url, 300)) ?? [];
}

export type ChartPoint = { t: number; price: number };

export async function getMarketChart(slug: string, days: number): Promise<ChartPoint[]> {
  const id = COIN_MAP[slug]?.id;
  if (!id) return [];
  const url = `${BASE}/coins/${id}/market_chart?vs_currency=usd&days=${days}`;
  const data = await getJson<{ prices: [number, number][] }>(url, 900);
  if (!data?.prices) return [];
  return data.prices.map(([t, price]) => ({ t, price }));
}

export type Global = {
  total_market_cap_usd: number;
  market_cap_change_24h: number;
  btc_dominance: number;
};

export async function getGlobal(): Promise<Global | null> {
  const data = await getJson<{
    data: {
      total_market_cap: { usd: number };
      market_cap_change_percentage_24h_usd: number;
      market_cap_percentage: { btc: number };
    };
  }>(`${BASE}/global`, 600);
  if (!data?.data) return null;
  return {
    total_market_cap_usd: data.data.total_market_cap.usd,
    market_cap_change_24h: data.data.market_cap_change_percentage_24h_usd,
    btc_dominance: data.data.market_cap_percentage.btc,
  };
}
